issue 5907 - making the navbar fixed

jaels opened the pull request:
Hey Julian,
Let me know what you think...
Thanks!

ghost commented:
I intended to specify that the navbar should be fixed to the top only when scrolling upwards after already scrolling down, until the point at which it meets the top of the page. For example, the way that the addressbar usually disappears and reappears when scrolling a mobile browser. 

This solution should also account for the position of the page content relative to the navbar. For example, if you run this branch locally and view http://localhost:5000/en/about then you will see that the first heading is below the navbar. This heading should appear in the same place that it does currently (https://adblockplus.org/en/about). And it should not change when the user scrolls down and back up - revealing the fixed navbar.

Does that make sense?

PS: I don't blame you for this misinterpretation. The ticket was not very clear 😞. Sorry about that.

jaels commented:
So, when the navbar appears again, when scrolling up, it shouldn't push the content down? that's what you mean?
(no worries, it did look a bit too simple...)

ghost commented:
I think so. The content should not move as a result of fixing or unfixing the navbar when scrolling up or down.

I think that this can be accomplished by positioning the navbar fixed when visible and adding the navbar height to the margin-top of the content.

We will probably also want to animate the fixed navbar in and out to create the illusion of scrolling it into view (again, like mobile browser address bars do).

This will require JavaScript. Therefore, we should make sure that whatever solution that we come up with also works / gracefully degrades without JavaScript.

jaels commented:
OK. Should I include the JS in main.js?

ghost commented:
Yes.

jaels commented:
Hey,
Hope it's what you meant.
Let me know what you think.

juliandoucette commented on static/css/main.css:
```diff
@@ -288,6 +288,16 @@ a.button.inverse
+
+#navbar.hide-navbar
+{
+    top: -4em;
```
NIT: The `#navbar` height is a min-height because `#navbar-menu` expands vertically on small screens. As a result, we cannot garantee that `top: -4em;` will hide the navbar (because min-height is not height).

Speaking of which, this seems to function strangely on mobile. If you open the menu and then scroll then the menu will remain and the navbar will hide. I guess it should stay in place on smaller screens (what do you think)?  And it seems this change somehow prevents us from scrolling the expanded list of languages in the menu 😕.

jaels replied:
It doesn't seem to expand in any of the mobile options in the dev tools, but yeah, if I minimize the screen extremely, I see that.
I agree that we can leave the navbar fixed in the smaller screens (is that what you mean?), so - max width 991px? smaller? That will solve also the -4em problem, won't it?

juliandoucette replied:
> It doesn't seem to expand in any of the mobile options in the dev tools, but yeah, if I minimize the screen extremely, I see that.

1. Shrink window down to mobile phone size
2. Expand navbar menu
3. Expand locale menu in navbar

> I agree that we can leave the navbar fixed in the smaller screens (is that what you mean?), so - max width 991px? smaller? That will solve also the -4em problem, won't it?

- Yes, leave the navbar visible (fixed) on smaller screens
- Yes, max-width 991
- No, that will not solve the -4em problem that I mentioned above
  - (Solving the -4em problem is not necessary to resolve this issue)
  - (To solve the -4em problem, one must put everything in the navbar except the navbar menu (currently only the navbar logo and the menu toggle button/link) into a transparent container that has a fixed height and apply the same fixed height to the navbar when the window is more than 992px wide. After that, the min-height will no longer be necessary. Does that make sense?)

juliandoucette commented on static/css/index-mobile.css:
```diff
@@ -1,7 +1,6 @@
 #content
 {
   text-align: center;
-  padding-top: 2.5em;
```
I'm still seeing the navbar overlap the first heading e.g. http://localhost:5000/en/about. The body content should be in the same place as it is currently e.g. https://adblockplus.org/en/about.

juliandoucette replied:
The overlap is gone. But the spacing above the main content is still not the same as it was before this change. Please try to retain the spacing (or ask for help if it's not trivial / you're not up for it).

jaels commented on static/js/main.js:
```diff
@@ -1,33 +1,62 @@
-
-  initLanguageSelection();
-  initMenu();
+(function() {
```
About your first comment - no problem, I actually really appreciate that. 
And about this one - how do I actually use the eslint and stylelint in this project? I followed the instructions, but I don't see that it works (installed it globally, trying to use it on a specific file etc)

juliandoucette replied:
Good question. Will you take this one @ireade? 

Anyway, you may ignore this issue for now. On second thought, I'm not sure that it makes sense to nit-pick contributors about coding style issues. We can probably prettify your code to match our coding style when we merge it.

ireade replied:
We are currently working on getting our `stylelint-config-eyeo` on npm, so unfortunately right now it can't be installed globally. 

However, you can install/configure it by manually downloading the repo and referencing it in your `stylelintrc.json` file: 

```json
{
  "extends": "path/to/codingtools/stylelint-config-eyeo"
}
```

But as Julian said, you can probably ignore the styling issues for now / let us handle it.

jaels commented on static/js/main.js:
```diff
@@ -1,33 +1,62 @@
+            .classList.toggle("visible");
+    }
+
+    // Hide navbar on scroll down, show on scroll up
```
OK, I'll try

jaels commented:
Hey!
I narrowed and organised the code, and addressed what you asked for.
Let me know what you think. 
Cheers

jaels commented:
Hey,
I'm will make another PR sometime soon, so better not to spend time on this one

jaels commented:
I think I addressed all the issues, hope it's getting us closer. Please ignore the previous commit - there were some problems with it.
Let me know what you think...
Cheers

juliandoucette commented on static/js/main.js:
```diff
@@ -1,33 +1,59 @@
+            .addEventListener("click", navigationClick, false);
+    }
+
+    function navbarToggle() {
```
NIT/Suggest: `initNavbarToggle` to be consistent with the other `init` functions in this file.

juliandoucette commented on static/js/main.js:
```diff
@@ -1,33 +1,59 @@
+
+    function navbarToggle() {
+        var navBar = document.getElementById("navbar");
+        var navbarHeight = navBar.offsetHeight;
```
According to MDN "This property will return null on Internet Explorer (9) if the style.position of the element itself is set to "fixed""... We may gracefully degrade (not support this feature without causing an error or other unexpected behaviour) this feature in IE9 if there is not another painless way to get offsetHeight across browsers including IE9.

This could be done by e.g. `if (!navbarHeight) return;`. (if we add a comment above explaining why this is necessary).

juliandoucette commented on static/js/main.js:
```diff
@@ -1,33 +1,59 @@
+        var lastScrollTop = 0;
+
+        window.addEventListener("scroll", (function() {
+            scrollHandled = true;
```
NIT: "handled" usually means "taken care of". AFAICT (As Far As I Can Tell) the purpose of setting this boolean in this function is to flag something "to be handled" and not "as taken care of". Therefore, setting "scrollHandled" (which I interpret as "has the scroll been handled"?) to "true" ("yes!") is misleading IMO.

I suggest setting it to false here and checking `!scrollHandled` below instead of `scrollHandled` and returning `true` from `handleScroll` instead of `false`.

juliandoucette commented on static/js/main.js:
```diff
@@ -1,33 +1,59 @@
+        }));
+
+        setInterval(function() {
+            if(window.innerWidth>991) {
```
NIT/Suggest: Use a constant (actually a variable in ES5; unfortunately) for 991 instead of hard coding this value to make it clearer what you're comparing to. e.g. `if (window.innerWidth > desktopBreakpoint)`

juliandoucette commented on static/js/main.js:
```diff
@@ -1,33 +1,59 @@
+                    scrollHandled = handleScroll();
+                }
+            }
+        }, 250);
```
TOL: (Why did you choose 250?)

jaels commented:
Hey Julian,
New version, I think we're close...
offsetHeight in IE9 - couldn't find a painless solution for that :( 
setInterval 250 - I answered it before, but it was probably lost in all the PRs and responses. 
No specific reason, I just saw it in some online examples of handling scroll events. 
Can be changed, of course, what do you think?

Waiting for your feedback,
Good day!
Yael

juliandoucette commented on static/js/main.js:
```diff
@@ -1,33 +1,64 @@
+            .addEventListener("click", navigationClick, false);
+    }
+
+    function initNavbarToggle() {
```
Bug: If you scroll down to hide the navbar and then resize the window < 991px then the navbar will not re-appear.

juliandoucette commented on static/js/main.js:
```diff
@@ -1,33 +1,64 @@
+            .addEventListener("click", navigationClick, false);
+    }
+
+    function initNavbarToggle() {
```
Bug: If you open the locale menu in the navbar and then scroll down then the navbar will disappear; leaving the open menu behind.

juliandoucette replied:
I forgot to mention that when one clicks the locale menu they are automatically scrolled to the top of the page. This is because the locale menu href is "#". We should probably remove this href and apply the appropriate cursor to this link/button via CSS in order to avoid this undesirable effect (auto-scrolling to the top).

jaels commented:
about the resize bug - is it ok that when I'm on a big screen, and then resize < 991, it will keep the toggle activity? so not become fixed (and the other way around), assuming not a lot of people resize, and probably none of them resize to a mobile screen. 

locale menu - so, when I scroll down it all disappears, and when I scroll up it reappears with the open locale? (assuming it was open before)

ghost commented:
> about the resize bug - is it ok that when I'm on a big screen, and then resize < 991, it will keep the toggle activity? 

I think that the navbar should show when the user resizes the window below the desktop breakpoint (and when they resize their window above the desktop breakpoint until they scroll).

> so not become fixed (and the other way around), assuming not a lot of people resize, and probably none of them resize to a mobile screen.

I would not make this assumption. IIRC the average laptop screen width in 2018 is 1366 and half of 1366 (683) is less than 991. And 991/1366 = ~0.725 (991 is 72% width of 1366). Therefore, an average user that does not maximize their window may see the mobile website. And an average user that snaps their window to half of their screen width will probably see the mobile website.

> locale menu - so, when I scroll down it all disappears, and when I scroll up it reappears with the open locale? (assuming it was open before)

Good question. I don't know. I guess I don't want to hide the locale menu on scroll because one may scroll by accident while they are trying to select a locale. Therefore, I think it makes more sense to not hide the navbar while the locale menu is open.

ghost commented:
Heads up: We're in the process of moving websites issue from trac to gitlab. I've left this issue open on trac (because you're working on it). Please see https://gitlab.com/eyeo/ and https://gitlab.com/eyeo/web.adblockplus.org for future websites related issues. I will also link from trac to gitlab as much as possible whenever closing or migrating issues.

jaels commented:
I did what you asked about the href="#" (apart from the rest), and now it doesnt scroll to the top, but didn't understand what you meant about the cursor - as far as I can see it acts the same as it did before. 
Let me know...

ghost commented:
> but didn't understand what you meant about the cursor - as far as I can see it acts the same as it did before.

That is because the `cursor: pointer;` style is provided by defaults.css. Otherwise, removing the `href` from an anchor tag would change the cursor to normal.

(Sorry about the confusion, I forgot that I covered this in website-defaults [defaults.css].)

juliandoucette commented on static/js/main.js:
```diff
@@ -1,33 +1,68 @@
+    function initNavbarToggle() {
+        var navBar = document.getElementById("navbar");
+        var navBarLocale = document.getElementById("navbar-locale-menu");
+        var navbarHeight = navBar.offsetHeight;
```
NIT: inconsistent capitalization of "navbar". I prefer "navbar" to "navBar", but I don't mind if you choose "navBar".

juliandoucette commented on static/js/main.js:
```diff
@@ -1,33 +1,68 @@
+
+        setInterval(function() {
+            if(window.innerWidth > desktopBreakpoint) {
+                if (!scrollHandled && navBarLocale.className != "visible" ) {
```
NIT/Suggest: Use `navBarLocale.classList.contains("visible")` because `navBarLocale.className` may contain other classes that falsify comparison to "visible" in the future.

(If I add another unreleased class to this element then this className comparison will never be truthy.)

juliandoucette commented on static/js/main.js:
```diff
@@ -1,33 +1,68 @@
+            scrollHandled = false;
+        }));
+
+        setInterval(function() {
```
I think that this solution is clever (in the good way) because it doesn't rely on the window resize event.

jaels commented:
Done.
BTW, not sure if you noticed (or maybe it's an existing issue) - the locale menu is cut at the bottom

ghost commented:
> the locale menu is cut at the bottom

I don't know what you mean?

jaels commented:
![screen shot 2018-02-27 at 12 21 43](https://user-images.githubusercontent.com/19172728/36726272-74bb3360-1bb9-11e8-8037-abe985c125e5.png)

jaels commented:
yeah, not really cut caus' it's scrollable, still could be cut in a more exact way, I guess

ghost commented:
> yeah, not really cut caus' it's scrollable, still could be cut in a more exact way, I guess

Ah, yes. This is intentional (although not optimal). We will refactor this menu later.

After you brought this issue to my attention, I discovered another caused by this change. Now that the navbar is fixed, the menu is not scrollable on mobile devices when the locale section is expanded. I'm not sure how to best address this issue yet. You may try to address it if you wish. Otherwise I will address it before pushing your change.

Detail: In order to make the menu scrollable on mobile, a new or existing parent of the menu must have a restricted height and `overflow-y` of `scroll`. e.g. `#navbar-menu {max-height: 80vh; overflow: scroll;}` (example only, `vh` units are not supported until IE11. However, using `vh` with a fallback to a fixed `px` value e.g. `max-height: 416px; max-height: 80vh;` would be acceptable for now... 416 is 480 (smallest mobile device commonly supported) minus 64 (the height of the navbar logo)).

ghost commented:
Sorry for the delay. I hadn't noticed that you fixed the scroll bug 😄. This change has now been merged/squashed/imported into our repository. Well done!! 🎉And thank you 🙏!!

Mercurial: https://hg.adblockplus.org/web.adblockplus.org/rev/84ac72993b9d

Github: https://github.com/adblockplus/web.adblockplus.org/commit/319a168a9afb33cbd197b47aedb7ba1236251297
